third_party/txt/src/txt/paragraph.cc

```diff
@@ -545,17 +545,23 @@ void Paragraph::Layout(double width, bool force) {
           int32_t next_grapheme;
           if (run.is_rtl()) {
             next_grapheme = grapheme_breaker_->previous();
+            if (next_grapheme == icu::BreakIterator::DONE)
+              break;
             subglyph_code_unit_counts.push_back(current_grapheme -
                                                 next_grapheme);
           } else {
             next_grapheme = grapheme_breaker_->next();
+            if (next_grapheme == icu::BreakIterator::DONE)
+              break;
             subglyph_code_unit_counts.push_back(next_grapheme -
                                                 current_grapheme);
             while (next_grapheme < static_cast<int32_t>(text_count)) {
               if (layout.getCharAdvance(next_grapheme) != 0)
                 break;
-              size_t sub_grapheme = grapheme_breaker_->current();
-              next_grapheme = grapheme_breaker_->next();
+              if (grapheme_breaker_->next() == icu::BreakIterator::DONE)
+                break;
+              size_t sub_grapheme = next_grapheme;
+              next_grapheme = grapheme_breaker_->current();
```

Comment:
Is this switch intentional (or maybe I'm missing something)? On the next line, it looks like we used to compute `grapheme_breaker_->next()` - `grapheme_breaker_->current()` but now we do `grapheme_breaker_->current() - grapheme_breaker_->next()`.

Reply:
This was an attempt to made this simpler.  It calls grapheme_breaker_->next() to advance the grapheme iterator.  If that fails, it aborts.

If it succeeds, then next_grapheme is updated to the new current position of the iterator, and the count of code units is the difference between the new next_grapheme and the old next_grapheme.

Reply:
Ah! Thanks - my brain skipped over the `->next()` call in the check on l. 561 and read it the same as the other additions above! Makes perfect sense.